File: packages/android/core-domain/src/main/kotlin/app/notes/domain/AIRouting.kt
package app.notes.domain

/**
 * Canonical AI task taxonomy. Matches the TS twin and the keys M09's
 * V1_ROUTES already uses.
 */
enum class AITask(val key: String) {
    Chat("chat"),
    Summarize("summarize"),
    SuggestTags("suggest_tags"),
    SuggestTitle("suggest_title"),
    ExtractActions("extract_actions"),
    Rewrite("rewrite"),
    Briefing("briefing"),
    Compression("compression"),
    Embed("embed"),
    ;

    companion object {
        fun fromKey(key: String): AITask? = entries.firstOrNull { it.key == key }
    }
}

data class AdapterBinding(val provider: AIProviderId, val model: String) {
    init { require(model.isNotBlank()) { "AdapterBinding.model cannot be empty" } }
}

/**
 * Partial routing override. Any null entry resolves to V1_ROUTES at
 * routing time.
 */
data class RoutingConfig(
    val chat: AdapterBinding? = null,
    val summarize: AdapterBinding? = null,
    val suggestTags: AdapterBinding? = null,
    val suggestTitle: AdapterBinding? = null,
    val extractActions: AdapterBinding? = null,
    val rewrite: AdapterBinding? = null,
    val briefing: AdapterBinding? = null,
    val compression: AdapterBinding? = null,
    val embed: AdapterBinding? = null,
) {
    fun bindingFor(task: AITask): AdapterBinding? = when (task) {
        AITask.Chat -> chat
        AITask.Summarize -> summarize
        AITask.SuggestTags -> suggestTags
        AITask.SuggestTitle -> suggestTitle
        AITask.ExtractActions -> extractActions
        AITask.Rewrite -> rewrite
        AITask.Briefing -> briefing
        AITask.Compression -> compression
        AITask.Embed -> embed
    }
}

/** Provider-keyed endpoint map. Only OLLAMA has a meaningful entry in V1. */
typealias ProviderEndpoints = Map<AIProviderId, String>

/**
 * Full ai_prefs payload. version lets the persisted shape evolve.
 */
data class AIPrefsPayload(
    val version: Int = 1,
    val routing: RoutingConfig? = null,
    val endpoints: ProviderEndpoints = emptyMap(),
)

/** M15 settings-side port. The Android adapter binds via Hilt. */
interface AIRouting {
    suspend fun providersAvailable(): List<AIProviderId>
    suspend fun getRouting(): RoutingConfig
    suspend fun setRouting(config: RoutingConfig)
    suspend fun getEndpoint(provider: AIProviderId): String?
    suspend fun setEndpoint(provider: AIProviderId, url: String)
    suspend fun resolveFor(task: AITask): AdapterBinding
}

/**
 * V2 key store. Lights up when Anthropic / OpenAI providers land —
 * V1 ships the table + RLS only.
 */
interface AIKeyStore {
    suspend fun putKey(provider: AIProviderId, secret: String)
    suspend fun hasKey(provider: AIProviderId): Boolean
    suspend fun deleteKey(provider: AIProviderId)
}

File: packages/domain/src/ai-routing.ts
import type { AIProviderId } from "./enums.js";

/**
 * Canonical task taxonomy for AI routing. Matches the keys M09's
 * `V1_ROUTES` already uses, so the M15 dynamic router replaces the
 * static one without churning consumers.
 */
export type AITask =
  | "chat"
  | "summarize"
  | "suggest_tags"
  | "suggest_title"
  | "extract_actions"
  | "rewrite"
  | "briefing"
  | "compression"
  | "embed";

export const AI_TASKS: readonly AITask[] = [
  "chat",
  "summarize",
  "suggest_tags",
  "suggest_title",
  "extract_actions",
  "rewrite",
  "briefing",
  "compression",
  "embed",
] as const;

/** A single (provider, model) pair. Provider is the canonical enum. */
export interface AdapterBinding {
  readonly provider: AIProviderId;
  readonly model: string;
}

/**
 * The persisted shape under `users_profile.ai_prefs.routing`. All task
 * keys are optional — missing entries fall back to V1_ROUTES at resolve
 * time so partial overrides are safe to ship.
 */
export interface RoutingConfig {
  readonly chat?: AdapterBinding;
  readonly summarize?: AdapterBinding;
  readonly suggest_tags?: AdapterBinding;
  readonly suggest_title?: AdapterBinding;
  readonly extract_actions?: AdapterBinding;
  readonly rewrite?: AdapterBinding;
  readonly briefing?: AdapterBinding;
  readonly compression?: AdapterBinding;
  readonly embed?: AdapterBinding;
}

/**
 * Endpoint URL per provider, indexed by provider id. Ollama is the only
 * one with a meaningful endpoint in V1 (well-known URLs cover OpenAI /
 * Anthropic in V2). Empty / missing entry → fall back to the
 * `OLLAMA_ENDPOINT_URL` env var that already serves M09.
 */
export type ProviderEndpoints = Partial<Record<AIProviderId, string>>;

/**
 * Full payload stored under `users_profile.ai_prefs`. `version` lets us
 * migrate the shape later; routing + endpoints are independent so a
 * routing-only update doesn't reset the endpoint.
 */
export interface AIPrefsPayload {
  readonly version: 1;
  readonly routing?: RoutingConfig;
  readonly endpoints?: ProviderEndpoints;
}

/**
 * The settings-side port. The M15 adapter (web + Android) reads / writes
 * this; M09's router reads through it to pick the active route.
 */
export interface AIRouting {
  providersAvailable(): Promise<readonly AIProviderId[]>;
  getRouting(): Promise<RoutingConfig>;
  setRouting(config: RoutingConfig): Promise<void>;
  getEndpoint(provider: AIProviderId): Promise<string | null>;
  setEndpoint(provider: AIProviderId, url: string): Promise<void>;
  resolveFor(task: AITask): Promise<AdapterBinding>;
}

/**
 * V2 key store. Built in V1 (table exists, RLS enforced) but unused —
 * lights up when Anthropic / OpenAI providers land.
 */
export interface AIKeyStore {
  putKey(provider: AIProviderId, secret: string): Promise<void>;
  hasKey(provider: AIProviderId): Promise<boolean>;
  deleteKey(provider: AIProviderId): Promise<void>;
}
